packages/easysearch:core/lib/core/search-collection.js

```diff
@@ -201,17 +201,44 @@ class SearchCollection {
         resultsHandle && resultsHandle.stop();
       });
 
+      let observedDocs = [];
+
+      const updateDocWithCustomFields = (doc, sortPosition) => collectionScope
+        .addCustomFields(doc, {
+          originalId: doc._id,
+          sortPosition,
+          searchDefinition: definitionString,
+          searchOptions: optionsString,
+        });
+
       let resultsHandle = cursor.mongoCursor.observe({
         addedAt: (doc, atIndex, before) => {
           doc = collectionScope.engine.config.beforePublish('addedAt', doc, atIndex, before);
-          doc = collectionScope.addCustomFields(doc, {
-            searchDefinition: definitionString,
-            searchOptions: optionsString,
-            sortPosition: atIndex,
-            originalId: doc._id
-          });
+          doc = updateDocWithCustomFields(doc, atIndex);
 
           this.added(collectionName, collectionScope.generateId(doc), doc);
+
+          // reorder all observed docs to keep valid sorting
+          observedDocs = observedDocs.map((doc, docIndex) => {
```

Comment:
@matteodem Wouldn't this be run for all the docs added when the client connects? does it make sense to do an initialization check here, as in this SO [answer](http://stackoverflow.com/questions/21355802/meteor-observe-changes-added-callback-on-server-fires-on-all-item)?